sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -247,14 +253,7 @@ class RocksDB(
   }
 
   def prefixScan(prefix: Array[Byte]): Iterator[ByteArrayPair] = {
-    val threadId = Thread.currentThread().getId
-    val iter = prefixScanReuseIter.computeIfAbsent(threadId, tid => {
```

Comment:
I noticed that the iterator reuse logic is removed. Is it related to the change? This could get CPU regression.

Reply:
My memory is broken but there should be a reason I didn't follow the cleanup approach of iterator in iterator(). That relies on task completion listener, and I roughly remember there was an issue with iterator in prefixScan so I couldn't do the same. 

If that was related to WriteBatch then we may not need this anymore and just follow the approach of iterator(), but honestly I didn't track down the issue as such.

That said, if we remove this, we probably need to employ the cleanup approach for we use in iterator().

Reply:
Yea so the previous model assumed that the iterator could reflect changes to writeBatch for the given thread. That is not true with the change any more.

So basically the following would fail, if we were to use cached iterator.
- Load version 0
- Add kv pairs (c, 1) (c, 2) (c, 3)
- commit 0
- Load version 1
- Call prefix scan and load map
- Add kv pairs (c, 4) (c, 5) (c, 6)
- Read key-values and now if we rely only on the old iterator loaded in the map, we read (c, 1), (c, 2) (c, 3). With the old writeBatch in place, the iterator was valid for the writeBatch as well which is why we could read all the kv pairs correctly.

Sample failure:
```
[info]   Set((("c", 1), 1), (("c", 2), 2), (("c", 3), 3)) did not equal Set((("c", 4), 4), (("c", 3), 3), (("c", 1), 1), (("c", 5), 5), (("c", 2), 2), (("c", 6), 6)) (StateStoreSuite.scala:917)
[info]   Analysis:
[info]   Set(missingInLeft: [(("c", 4), 4), (("c", 5), 5), (("c", 6), 6)])
```

Reply:
It is simpler than that - cached prefixScan iterator does not live across microbatches. We close it in both commit and rollback. So it's not really a cache but closer to reference tracking.

Reply:
Yes correct. we close it in commit & rollback. In this case, the call is made twice though. Basically to check values in previous batch and then newly added values in current batch.

Reply:
My bad, you're right I forgot about it. There are two usages for prefix scan, and I intentionally didn't close the iterator per each usage to reuse the underlying iterator among two usages. But that optimization seems to be dangerous as it will lead to correctness issue if there is any possibility for two usages to use the iterator at the same time. If the benefit we can get from dangerous optimization is not huge, I'd lean on removing the optimization. Just that we need to revisit the resource cleanup for prefix scan iterator then.

Reply:
Yea use-case seems quite narrow and might not be safe if its passed around too. I would probably prefer to remove this. Lmk what you think ?

Reply:
Done - added the close and cleanup logic to mimic iterator()

Reply:
I would follow your intuition that removing the optimization won't cause a big issue. I guess if we would keep it, the logic might be more complicated, so it makes sense, so it makes sense.